src/app/clusters/bindings/BindingManager.cpp

```diff
@@ -0,0 +1,192 @@
+{
+    BindingManager * manager = static_cast<BindingManager *>(context);
+    manager->HandleDeviceConnected(device);
+}
+
+void BindingManager::HandleDeviceConnected(OperationalDeviceProxy * device)
+{
+    mPendingClusterMap.ForEachActiveObject([&](PendingClusterEntry * entry) -> Loop {
+        if (entry->GetPeerId() == device->GetPeerId())
+        {
+            SyncPendingClustersToPeer(device, entry);
+        }
+
+        return Loop::Continue;
+    });
+}
+
+void BindingManager::SyncPendingClustersToPeer(OperationalDeviceProxy * device, PendingClusterEntry * pendingClusters)
+{
+    for (const ClusterPath & path : *pendingClusters)
+    {
+        ClusterId cluster   = path.cluster;
+        EndpointId endpoint = path.endpoint;
+        for (uint8_t j = 0; j < EMBER_BINDING_TABLE_SIZE; j++)
+        {
+            EmberBindingTableEntry entry;
+            if (emberGetBinding(j, &entry) == EMBER_SUCCESS && entry.type != EMBER_UNUSED_BINDING && entry.clusterId == cluster &&
+                entry.local == endpoint && mBoundDeviceChangedHandler)
+            {
+                mBoundDeviceChangedHandler(entry.local, entry.remote, cluster, device);
+            }
+        }
+    }
+    mPendingClusterMap.RemoveEntry(pendingClusters);
+}
+
+void BindingManager::HandleDeviceConnectionFailure(void * context, PeerId peerId, CHIP_ERROR error)
+{
+    BindingManager * manager = static_cast<BindingManager *>(context);
+    manager->HandleDeviceConnectionFailure(peerId, error);
+}
+
+void BindingManager::HandleDeviceConnectionFailure(PeerId peerId, CHIP_ERROR error)
+{
+    // Simply release the entry, the connection will be re-established on need.
+    ChipLogError(AppServer, "Failed to establish connection to node 0x" ChipLogFormatX64, ChipLogValueX64(peerId.GetNodeId()));
+    mAppServer->GetCASESessionManager()->ReleaseSession(peerId);
+}
+
+CHIP_ERROR BindingManager::DisconnectDevice(FabricIndex fabric, NodeId node)
+{
+    FabricInfo * fabricInfo     = mAppServer->GetFabricTable().FindFabricWithIndex(fabric);
+    PeerId peer                 = fabricInfo->GetPeerIdForNode(node);
+    PendingClusterEntry * entry = mPendingClusterMap.FindEntry(peer);
+
+    VerifyOrReturnError(entry != nullptr, CHIP_ERROR_NOT_FOUND);
+
+    mAppServer->GetCASESessionManager()->ReleaseSession(peer);
+    mPendingClusterMap.RemoveEntry(entry);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingManager::NotifyBoundClusterChanged(EndpointId endpoint, ClusterId cluster)
+{
+    for (uint8_t i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
+    {
+        EmberBindingTableEntry entry;
+
+        if (emberGetBinding(i, &entry) == EMBER_SUCCESS && entry.type != EMBER_UNUSED_BINDING && entry.local == endpoint &&
+            entry.clusterId == cluster)
+        {
+            FabricInfo * fabricInfo             = mAppServer->GetFabricTable().FindFabricWithIndex(entry.fabricIndex);
+            PeerId peer                         = fabricInfo->GetPeerIdForNode(entry.nodeId);
```

Comment:
nullptr check for fabricinfo here as well. Also at method top validate mAppServer